decomp/semantics/uds.py

```diff
@@ -985,12 +1079,178 @@ def document_id(self) -> str:
         syntax_root = self.get_digraph_root(self.syntax_subgraph)
         return self.syntax_nodes[syntax_root]['document_id']
 
+    # TODO: track at the document level
     @memoized_property
     def sentence_id(self) -> str:
         """Sentence ID from Universal Dependencies"""
         syntax_root = self.get_digraph_root(self.syntax_subgraph)
         return self.syntax_nodes[syntax_root]['sentence_id']
 
+class UDSDocumentGraph(UDSGraph):
+    """A Universal Decompositional Semantics document-level graph
+
+    Parameters
+    ----------
+    graph
+        the NetworkX DiGraph from which the document-level graph
+        is to be constructed
+    name
+        the name of the graph
+    """
+    @overrides
+    def __init__(self, graph: DiGraph, name: str):
+        super().__init__(graph, name)
+
+    # TODO: support querying?
+
+    def add_annotation(self,
+                   node_attrs: Dict[str, Dict[str, Any]],
+                   edge_attrs: Dict[str, Dict[str, Any]]) -> None:
+        """Add node and or edge annotations to the graph
+
+        Parameters
+        ----------
+        node_attrs
+            the node annotations to be added
+        edge_attrs
+            the edge annotations to be added
+        """
+        for node, attrs in node_attrs.items():
+            self._add_node_annotation(node, attrs)
+
+        for edge, attrs in edge_attrs.items():
+            self._add_edge_annotation(edge, attrs)
+
+    def _add_edge_annotation(self, edge, attrs):
+        if edge in self.graph.edges:
+            self.graph.edges[edge].update(attrs)
+        else:
+            self.graph.add_edges_from([(edge[0], edge[1], attrs)])
+
+    def _add_node_annotation(self, node, attrs):
+        # We do not currently have a use case for document node annotations,
+        # but it is included for completeness.
```

Comment:
Would be curious to know if you foresee a need for this.

Reply:
One reason I could see wanting this is in case we wanted to implement coref annotations as node attributes rather than edge attributes. For instance, if a `UDSDocument` had an `entities` attribute that was a set of entity identifiers, and we had node attributes that stored an identifier. Otherwise, it's not super clear to me what else node attributes might store. It sort of feels like most of the time the sort of things we would want to store at the document level are relational in nature and so they're best conceptualized as edge attributes.

Reply:
That makes sense. I'll leave it in, then, with a comment about the possible use case.

Reply:
It might be worth raising a warning if someone tries to use this method on `UDSDocumentGraph`s.